platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -17,68 +19,374 @@ components within the platform.
+current address book. The previous roster will continue to be used to validate events that have an older software
+version. The switch to using event birth rounds will happen in a future proposal for Dynamic Address Book phase 3.
+
+During the handling of a freeze upgrade transaction, instead of writing the new config.txt to disk, a new candidate
+roster is created and stored in the roster merkle state.
+
+NOTE: The roster state maintains its own `Roster History`. Setting the active roster in the roster state will cause
+the history to be updated by rotating the current active roster to the previous active roster and setting the new
+roster as the current active roster. The old previous active roster is removed from the history and the merkle state.
+
+![StartupAndRosters-TSS-Roster.drawio.svg](StartupAndRosters-TSS-Roster.drawio.svg)
+
+#### Startup: Determining The New Roster History
+
+There are four distinct modes that a node can start in:
+
+* `Genesis Network` - The node is started with a genesis roster and no pre-existing state on disk.
+* `Network Transplant` - The node is started with a state on disk and an overriding roster for a different network.
+* `Software Upgrade` - The node is restarted with the same state on disk and a software upgrade is happening.
+* `Normal Restart` - The node is restarted with the same state on disk and no software upgrade is happening.
+
+A decision tree for deciding which mode to start the network in is as follows:
+
+![](TSS%20Roster%20Lifecycle-Proposed%20Startup%20Behavior.drawio.svg)
+
+##### MODE: New Network Genesis
+
+Conditions:
+
+1. No state loaded from disk
+2. A `genesis-config.txt` file is present on disk
+
+Ignored Files: (Log warning if present)
+
+1. `config.txt`
+2. `override-config.txt`
+
+Startup Tasks:
+
+1. Translate `genesis-config.txt` to a genesisRoster.
+2. Create a genesis state with the genesisRoster.
+3. Set (genesisRoster, 0) ase the new active roster in the roster state.
+4. rosterHistory := `[(genesisRoster, 0)]`
+5. Start the platform with the rosterHistory.
+
+Cleanup Tasks:
+
+1. Once the first saved state is created, All `config.txt` related files are moved to subdirectory
+   `.archive/yyyy-MM-dd_HH-mm-ss/`.
+
+##### MODE: Migration Software Upgrade
+
+Conditions:
+
+1. A state is loaded from disk
+2. A software upgrade is happening
+3. A `config.txt` file is present on disk
+4. The roster state is empty (no candidate roster and no active rosters)
+5. No `override-config.txt` file is present on disk
+
+Ignored Files: (Log warning if present)
+
+1. `genesis-config.txt`
+
+Startup Tasks:
+
+1. Read the current AddressBooks from the platform state.
+2. configAddressBook := Read the address book in config.txt
+3. previousRoster := translateToRoster(currentAddressBook)
+4. currentRoster := translateToRoster(configAddressBook)
+5. currentRound := state round +1
+6. set (previousRoster, 0) as the active roster in the roster state.
```

Comment:
Why the round number is 0? Shouldn't it be `currentRound - 1` which equals the `state round`?

Reply:
The full answer requires some context.  As part of our initial Dynamic Address Book effort, where we want to support the ability to switch address books (rosters) without a system restart, we planned to use an event's `birthRound` as the means of determining which roster to use to find the public key to validate the signature on the event.  The `birthRound` is the latest consensus round number when the event is created.   The rosters are associated with the round number that it becomes active and governs consensus.  We use the roster with the highest round in becoming active that is below or equal to the event's birth round to validate the event's signature.   

As of right now, we have not completed the switch to using `birthRound` in this process.  The old code doesn't store round numbers and uses the address book by position to determine "current" or "previous".  Right now we only update the roster at software upgrade boundaries.  After a software upgrade, the old events from the previous software version still need to be validated.  We use the previous roster for this.  In this migration, the round numbers do not matter, only the position / order that determined current and previous roster.   Going forward, when rosters become active in the state, they are associated with the round number they become active.  When we switch to fully dynamic address books / dynamic rosters, the round numbers will matter and we will have accurate round numbers there. 

I chose 0 because we don't know the actual number.  While it is somewhat arbitrary since we are selecting it by position, there will be events in the pre-consensus event stream that are older than `currentRound -1`.    We don't know how many rounds back the PCES will have events for.  0 gives a guarantee that it is prior to all possible events.

Reply:
The API on the roster state only allows setting the "current" active roster.   When two rosters are set, the first roster is rotated out to become the "previous" or second roster.   The instruction here to set (previous, 0) as the active roster is so that it can be pushed back into the second position when we set (currentRoster, currentRound) as the active roster.